README.md
# vite-plugin-multiple

Allow multiple Vite to run simultaneously.

[![NPM version](https://img.shields.io/npm/v/vite-plugin-multiple.svg)](https://npmjs.com/package/vite-plugin-multiple)
[![NPM Downloads](https://img.shields.io/npm/dm/vite-plugin-multiple.svg)](https://npmjs.com/package/vite-plugin-multiple)

## Install

```sh
npm i -D vite-plugin-multiple
```

## Usage

```js
import multiple from 'vite-plugin-multiple'

export default {
  plugins: [
    multiple([
      {
        name: 'foo',
        config: 'vite.foo.config.mjs',
      },
      {
        name: 'bar',
        config: 'vite.bar.config.mjs',
      },
    ]),
  ],
}
```

**`vite serve`**

- `http://localhost:5173` access to the **main** app
- `http://localhost:5174` access to the **foo** app
- `http://localhost:5175` access to the **bar** app

**`vite build`**

- `dist` **main** app
- `dist/foo` **foo** app
- `dist/bar` **bar** app

## API <sub><sup>(Define)</sup></sub>

```ts
multiple(
  apps: {
    /**
     * Human friendly name of your entry point.
     */
    name: string
    /**
     * Vite config file path.
     */
    config: string
    /**
     * Explicitly specify the run command.
     */
    command?: 'build' | 'serve'
  }[],
  options: {
    /**
     * Called when all builds are complete.
     */
    callback?: () => void,
  } = {},
)
```


CHANGELOG.md
## 0.2.0 (2023-02-18)

- 221cb24 feat: support specify run command

## 0.1.1 (2023-01-17)

- 6b31cd6 fix: correct types

## 0.1.0 (2023-01-17)


test/build.test.ts
import fs from 'node:fs'
import path from 'node:path'
import { build as viteBuild } from 'vite'
import {
  beforeAll,
  afterAll,
  expect,
  test,
} from 'vitest'
import multiple from '../src'

const root = path.join(__dirname, 'fixtures')

beforeAll(async () => {
  await viteBuild({
    configFile: path.join(root, 'vite.config.mjs'),
    plugins: [multiple([
      {
        name: 'foo',
        config: path.join(root, 'foo/vite.config.mjs'),
      },
      {
        name: 'bar',
        config: path.join(root, 'bar/vite.config.mjs'),
      },
    ])],
  })
})

test('build', async () => {
  const { message: main } = await import('./fixtures/dist/main'!)
  const { message: foo } = await import('./fixtures/foo/dist/main'!)
  const { message: bar } = await import('./fixtures/bar/dist/main'!)
  expect(main).eq('main')
  expect(foo).eq('foo')
  expect(bar).eq('bar')
})

afterAll(() => {
  fs.rmSync(path.join(root, 'dist'), { recursive: true, force: true })
  fs.rmSync(path.join(root, 'foo/dist'), { recursive: true, force: true })
  fs.rmSync(path.join(root, 'bar/dist'), { recursive: true, force: true })
})


test/serve.test.ts
import path from 'node:path'
import {
  type ViteDevServer,
  createServer,
} from 'vite'
import {
  beforeAll,
  afterAll,
  expect,
  test,
} from 'vitest'
import fetch from 'node-fetch'
import multiple from '../src'

const root = path.join(__dirname, 'fixtures')
const servers: ViteDevServer[] = []

beforeAll(async () => new Promise(async resolve => {
  const server = await createServer({
    configFile: path.join(root, 'vite.config.mjs'),
    plugins: [multiple([
      {
        name: 'foo',
        config: path.join(root, 'foo/vite.config.mjs'),
      },
      {
        name: 'bar',
        config: path.join(root, 'bar/vite.config.mjs'),
      },
    ], {
      callback() {
        resolve()
      },
    })],
  })
  await server.listen()
}))

test('serve', async () => {
  const mainHTML = await (await fetch('http://localhost:5173')).text()
  const fooHTML = await (await fetch('http://localhost:5174')).text()
  const barHTML = await (await fetch('http://localhost:5175')).text()

  expect(mainHTML).includes('<h1>main</h1>')
  expect(fooHTML).includes('<h1>foo</h1>')
  expect(barHTML).includes('<h1>bar</h1>')
})

afterAll(async () => {
  for (const server of servers) {
    await server.close()
  }
  servers.length = 0
})


test/fixtures/main.ts
export const message = 'main'

vite.config.ts
import fs from 'node:fs'
import path from 'node:path'
import { spawn } from 'node:child_process'
import { builtinModules } from 'node:module'
import { defineConfig } from 'vite'

export default defineConfig({
  build: {
    emptyOutDir: false,
    minify: false,
    outDir: '',
    target: 'node14',
    lib: {
      entry: path.join(__dirname, 'src/index.ts'),
      formats: ['es', 'cjs'],
      fileName: format => format === 'cjs' ? '[name].cjs' : '[name].js',
    },
    rollupOptions: {
      external: [
        ...builtinModules
          .filter(m => !m.startsWith('_'))
          .map(m => [m, `node:${m}`])
          .flat(),
        'vite',
      ],
      output: {
        exports: 'named',
      },
    },
  },
})

function generateTypes() {
  return new Promise(resolve => {
    const cp = spawn(
      process.platform === 'win32' ? 'npm.cmd' : 'npm',
      ['run', 'types'],
      { stdio: 'inherit' },
    )
    cp.on('exit', code => {
      !code && console.log('[types]', 'declaration generated')
      resolve(code)
    })
    cp.on('error', process.exit)
  })
}

if (process.env.NODE_ENV !== 'test') {
  fs.rmSync('types', { recursive: true, force: true })
  generateTypes()
}


src/index.ts
import path from 'node:path'
import {
  type Plugin,
  type ResolvedConfig,
  type UserConfig,
  build as viteBuild,
  createServer,
  loadConfigFromFile,
  mergeConfig,
} from 'vite'

export type AppConfig = Parameters<typeof multiple>[0][number]

export default function multiple(
  apps: {
    /**
     * Human friendly name of your entry point.
     */
    name: string
    /**
     * Vite config file path.
     */
    config: string
    /**
     * Explicitly specify the run command.
     */
    command?: 'build' | 'serve'
  }[],
  options: {
    /**
     * Called when all builds are complete.
     */
    callback?: () => void,
  } = {},
): Plugin {
  let config: ResolvedConfig

  return {
    name: 'vite-plugin-multiple',
    config(config) {
      config.clearScreen ??= false
    },
    async configResolved(_config) {
      config = _config
    },
    configureServer(server) {
      if (server.httpServer) {
        server.httpServer.once('listening', () => run(config, apps, 'serve').then(() => options.callback?.()))
      } else {
        run(config, apps, 'serve').then(() => options.callback?.())
      }
    },
    async closeBundle() {
      if (config.command === 'build') {
        await run(config, apps, config.command)
        options.callback?.()
      }
    },
  }
}

export async function resolveConfig(config: ResolvedConfig, app: AppConfig): Promise<UserConfig> {
  const { config: userConfig } = (await loadConfigFromFile({
    command: app.command!,
    mode: config.mode,
    ssrBuild: !!config.build?.ssr,
  }, app.config)) ?? { path: '', config: {}, dependencies: [] };
  const defaultConfig: UserConfig = {
    root: config.root,
    mode: config.mode,
    build: {
      outDir: !userConfig.root || userConfig.root === /* conflict */config.root
        ? path.posix.join(config.build.outDir, app.name)
        : undefined,
    },
    clearScreen: false,
  }
  return mergeConfig(defaultConfig, userConfig);
}

export async function run(
  config: ResolvedConfig,
  apps: AppConfig[],
  mainAppCommand: ResolvedConfig['command'],
): Promise<void> {
  let port = 5174 // The port of main App is 5173

  for (const app of apps) {
    app.command ??= mainAppCommand

    const userConfig = await resolveConfig(config, app)

    if (app.command === 'serve') {
      userConfig.server ??= {}
      userConfig.server.port ??= port++

      const viteDevServer = await createServer({
        configFile: false,
        ...userConfig,
      })
      await viteDevServer.listen()
      viteDevServer.printUrls()
    } else {
      await viteBuild({
        // 🚧 Avoid recursive build caused by load default config file.
        configFile: false,
        ...userConfig,
      })
    }
  }
}
